lib/evaluate/check-call.cc

```diff
@@ -0,0 +1,199 @@
+        *kw);
+  }
+  if (auto type{arg.GetType()}) {
+    if (type->IsAssumedType()) {
+      messages.Say(
+          "Assumed type argument requires an explicit interface"_err_en_US);
+    } else if (type->IsPolymorphic()) {
+      messages.Say(
+          "Polymorphic argument requires an explicit interface"_err_en_US);
+    } else if (type->category() == TypeCategory::Derived) {
+      auto &derived{type->GetDerivedTypeSpec()};
+      if (!derived.parameters().empty()) {
+        messages.Say(
+            "Parameterized derived type argument requires an explicit interface"_err_en_US);
+      }
+    }
+  }
+  if (const auto *expr{arg.UnwrapExpr()}) {
+    if (auto named{ExtractNamedEntity(*expr)}) {
+      const semantics::Symbol &symbol{named->GetLastSymbol()};
+      if (symbol.Corank() > 0) {
+        messages.Say(
+            "Coarray argument requires an explicit interface"_err_en_US);
+      }
+      if (const auto *details{
+              symbol.detailsIf<semantics::ObjectEntityDetails>()}) {
+        if (details->IsAssumedRank()) {
+          messages.Say(
+              "Assumed rank argument requires an explicit interface"_err_en_US);
+        }
+      }
+      if (symbol.attrs().test(semantics::Attr::ASYNCHRONOUS)) {
+        messages.Say(
+            "ASYNCHRONOUS argument requires an explicit interface"_err_en_US);
+      }
+      if (symbol.attrs().test(semantics::Attr::VOLATILE)) {
+        messages.Say(
+            "VOLATILE argument requires an explicit interface"_err_en_US);
+      }
+    }
+  }
+}
+
+static void CheckExplicitInterfaceArg(const ActualArgument &arg,
+    const characteristics::DummyArgument &dummy, FoldingContext &context) {
+  std::visit(
+      common::visitors{
+          [&](const characteristics::DummyDataObject &object) {
+            if (const auto *expr{arg.UnwrapExpr()}) {
+              if (auto type{characteristics::GetTypeAndShape(*expr, context)}) {
+                object.type.IsCompatibleWith(context.messages(), *type);
+              } else {
+                // TODO
+              }
+            } else {
+              // TODO
+            }
+          },
+          [&](const characteristics::DummyProcedure &) {
+            // TODO check effective procedure compatibility
+          },
+          [&](const characteristics::AlternateReturn &) {
+            // TODO check alternate return
+          },
+      },
+      dummy.u);
+}
+
+static void RearrangeArguments(const characteristics::Procedure &proc,
+    ActualArguments &arg, parser::ContextualMessages &messages) {
+  CHECK(proc.HasExplicitInterface());
+  if (arg.size() < proc.dummyArguments.size()) {
+    arg.resize(proc.dummyArguments.size());
+  } else if (arg.size() > proc.dummyArguments.size()) {
+    messages.Say(
+        "%zd effective argument%s are passed in a reference to a procedure that expects only %zd"_err_en_US,
+        arg.size(), arg.size() == 1 ? "" : "s", proc.dummyArguments.size());
```

Comment:
A little suspect from the point of view of translatability.

Reply:
Yes, will reword.